spec/requests/orders_spec.rb

```diff
@@ -0,0 +1,140 @@
+        expect(response).to render_template "index"
+        expect(assigns[:orders]).to eq [order]
+      end
+    end
+  end
+
+  describe "POST #create" do
+    subject { post :create, params: { order: order_params } }
+
+    let(:cart_items) { { product.id => 1 }.to_json }
+    let(:order_params) { { address: "123 Main St", phone: "123-456-7890" } }
+
+    before { request.cookies[:cart_items] = cart_items }
+
+    after { clear_enqueued_jobs }
+
+    describe "callbacks" do
+      describe "#cart_items" do
+        it do
+          expect { subject }.to change { assigns(:cart_items) }
+                    .from(assigns(:cart_items)).to(JSON.parse(cart_items))
+        end
+      end
+    end
+
+    context "when valid product and order params" do
+      context "when create successfully" do
+        it do
+          expect { subject }.to change(Order, :count).by(1)
+          expect(response.cookies).to include("cart_items" => nil)
+          expect(response).to redirect_to orders_path
+          expect(flash[:success]).to eq "Order successfully placed! Please check your email to track the notification"  
+        end
+      end
+
+      context "when create order then sending mail place order is enqueued job " do
+        let(:order_mailer_jobs) do
+          ActiveJob::Base.queue_adapter.enqueued_jobs.select do |job|
+            job[:args][0] == "OrderMailer" && job[:args][1] == "place_order"
+          end
+        end
+
+        before { subject }
+
+        it { expect(order_mailer_jobs.size).to eq 1 }
+      end
+    end
+
+    context "when invalid product" do
+      let(:cart_items) { { "4" => 1 }.to_json }
+      let(:order_params) { { address: "123 Main St", phone: "123-456-7890" } }
+
+      it do
+        expect { subject }.not_to change(Order, :count)
+        expect(response).to redirect_to cart_path
+        expect(flash[:error]).to eq "Product not found."
+        expect(response.cookies).to include("cart_items" => "{}")
+      end
+    end
+
+    context "when invalid order params" do
+      let(:order_params) { { address: "123 Main St" } }
+
+      it do
+        expect { subject }.not_to change(Order, :count)
+        expect(flash[:error]).to eq "Order failed! Please fill out your address and phone"
+        expect(response).to redirect_to cart_path
+      end
+    end
+  end
+
+  describe "PATCH #update" do
+    subject { patch :update, params: params }
+
+    let(:params) { { id: order_id, order: { reason_description: "cancel this order", status: "cancel" } } }
+    let(:order) { create :order, user: user, status: status }
+    let(:status) { :wait_confirm }
+    let(:order_id) { order.id }
+
+    context "when order's status is wait_confirm" do
+      it do
+        expect { subject }.to change { order.reload.status }.from("wait_confirm").to("canceled")
+        expect(response).to redirect_to orders_path
+      end
+    end
+
+    context "when order's status in't wait_confirm" do
+      let(:status) { :delivering }
```

Comment:
**order's status in't wait_confirm** mà chỉ có status delivering à e?

Reply:
@tamtt-1031 dạ okie anh để em update lại ạ